### ISSUE TITLE
Improve predicability of Attr.TAGS and Scannable.tags() wrt duplicates, add a tagsDeduplicated method

### ISSUE THREAD

**Issue by simonbasle** [reactor/reactor-core#2542]
## Expected Behavior
Calling `.tags("1", "one").tag("1", "ein")` then scanning for the `Attr.TAGS` should output [`1`-`ein`].

## Actual Behavior
Scanning for the `Attr.TAGS` outputs both pairs: [`1`-`one`,`1`-`ein`].

`has-workaround` because effectively this is used by `FluxMetrics`, in which the tags are converted to a micrometer `Tags` instance which ends up deduplicating. The set iteration order is undefined technically though...

## Steps to Reproduce

```java
	@Test
	void overwritingTag() {
		final Flux<Object> source = Flux.empty().tag("1", "one");

		final Flux<Object> modified = source.tag("1", "ein");

		assertThat(Scannable.from(source).scan(Scannable.Attr.TAGS).map(t -> t.getT1()+t.getT2()))
				.containsExactly("1one");

		assertThat(Scannable.from(modified).scan(Scannable.Attr.TAGS).map(t -> t.getT1()+t.getT2()))
				.containsExactly("1ein");
	}
```

## Possible Solution
The internal storage and accumulation of tags in `FluxName` should use a `Map<String, String>`, probably one that guarantees iteration order like `LinkedHashMap`.

The `Set<Tuple2<String, String>>` abstraction is not adapted, despite having been used and exposed since 3.1.0.RELEASE. It is only exposed through the `Attr.TAGS` though, which unfortunately will have to stay like that for now. The conversion from `Map` can happen on demand inside `scanUnsafe`.

This change should target 3.3.x branch.

**Comment by simonbasle** [comment]
On one hand:
  - we expect the vast majority of users to only _set_ tags (via `Flux#tag(String, String)`) and then rely on `Flux#metrics()` to make use of these tags (including deduplication)
  - the "allow duplicates" aspect is never mentioned in the javadoc
  - we still think that the "map"-like no duplicates behavior would have been the least surprising, and that the set/duplicates one was a mistake on our part...

*BUT the fact remains that it it currently possible to observe duplicates, ie multiple tuples with same key but different values.*

As a result, we'll err on the conservative side and only perform that behavior change in 3.5.0.

**If you have an actual usecase relying on the duplicate-allowed aspect, please let us know until then*

**Comment by simonbasle** [comment]
There are two separate occurrences of this issue: 
 1. at the `FluxName` level (which currently stores `Tuple2<String, String>` inside a `Set`, which effectively allows different key-value pairs for a common key) => fixed by #2543 
 2. at the `Scannable.tags()` level (where distinct `FluxName` operators in the chain of parents might store different values for a common key, even if each `FluxName` itself internally solves 1. and deduplicates)

**Comment by simonbasle** [comment]
Let's embrace the duplicates _in storage_, but improve the predictability of iteration by using a `List` instead of a `Set`.
That way, `scan(Attr.TAGS)` will output duplicated pairs in declaration order.

`Scannable#tags()` becomes confusing when there are duplicates but also distinct underlying `FluxName` operators, i.e. there's a mix of contiguous `tag` operators and a break in the middle like so:

```java
Flux<Integer> f = Flux.just(1, 2, 3)
			.tag("key2", "earlyValue2")
			.tag("key1", "earlyValue1")
			.tag("key1", "earlyOverwriteValue1")
			.filter(i -> true) //this separates two macro-fused tag stages
			.tag("key1", "lateValue1")
			.tag("key2", "lateValue2")
			.tag("key1", "lateOverwriteValue1");
```
With the `List` approach, tags from each group of 3 macro-fused tags would appear in their declaration order but the downmost group would appear first:

```
lateValue1
lateValue2
lateOverwriteValue1
earlyValue2
earlyValue1
earlyOverwriteValue1
```

When streaming and iterating the `tags()`, the child > parent > grandparent order is not terribly useful I think. It is more reasonable to show most "recent" values last, e.g. if the goal is to add the pairs to a `Map` for deduplication for instance.

Since the deduplicated case is likely often desired, we can also add a `tagsDeduplicated()` method which returns a `Map<String, String>` of deduplicated tags. Keys in declaration order, values updated to reflect the one defined last in case of duplicates. With the above change, this can be done pretty fast:

```java
		return tags().collect(Collectors.toMap(Tuple2::getT1, Tuple2::getT2,
			(s1, s2) -> s2, LinkedHashMap::new));
```

So in summary:
1. `FluxName` and similar operators now store tags as `List<Tuple2<String, String>>` in declaration order in case of macro-fusion
2. (behavior change) `Scannable#tags` now starts from the uppermost reachable parent, and otherwise outputs tags as they are declared (including duplicates)
3. add a `Map<String, String> tagsDeduplicated()` method to `Scannable`

**Comment by simonbasle** [comment]
Here are a couple of tests to demonstrate the expected new behavior:

```java
	@Test
	void tagsIncludesDuplicatesAndReverseDeclarationOrder() {
		Flux<Integer> f = Flux.just(1, 2, 3)
			.tag("key2", "earlyValue2")
			.tag("key1", "earlyValue1")
			.tag("key1", "earlyOverwriteValue1")
			.filter(i -> true) //this separates two macro-fused tag stages
			.tag("key1", "lateValue1")
			.tag("key2", "lateValue2")
			.tag("key1", "lateOverwriteValue1");

		assertThat(Scannable.from(f).tags())
			.containsExactly(
				Tuples.of("key2", "earlyValue2"),
				Tuples.of("key1", "earlyValue1"),
				Tuples.of("key1", "earlyOverwriteValue1"),
				Tuples.of("key1", "lateValue1"),
				Tuples.of("key2", "lateValue2"),
				Tuples.of("key1", "lateOverwriteValue1")
			);
	}

	@Test
	void tagsDeduplicatedUsesLatestValueButOriginalKeyOrder() {
		Flux<Integer> f = Flux.just(1, 2, 3)
			.tag("key2", "earlyValue2")
			.tag("key1", "earlyValue1")
			.tag("key1", "earlyOverwriteValue1")
			.filter(i -> true) //this separates two macro-fused tag stages
			.tag("key1", "lateValue1")
			.tag("key2", "lateValue2")
			.tag("key1", "lateOverwriteValue1");

		assertThat(Scannable.from(f).tagsDeduplicated())
			.containsExactly(
				new AbstractMap.SimpleImmutableEntry<>("key2", "lateValue2"),
				new AbstractMap.SimpleImmutableEntry<>("key1", "lateOverwriteValue1")
			);
	}
```

**Comment by simonbasle** [comment]
I was also thinking about refreshing the documentation:

```java
	/**
	 * Visit this {@link Scannable} and its {@link #parents()}, starting by the furthest reachable parent,
	 * and return a {@link Stream} of the tags which includes duplicates and outputs tags in declaration order
	 * (grandparent tag(s) &gt; parent tag(s)  &gt; current tag(s)).
	 * <p>
	 * Tags can only be discovered until no parent can be inspected, which happens either
	 * when the source publisher has been reached or when a non-reactor intermediate operator
	 * is present in the parent chain (i.e. a stage that is not {@link Scannable} for {@link Attr#PARENT}).
	 *
	 * @return the stream of tags for this {@link Scannable} and its reachable parents, including duplicates
	 * @see #tagsDeduplicated()
	 */
	default Stream<Tuple2<String, String>> tags() {
		...
	}

	/**
	 * Visit this {@link Scannable} and its {@link #parents()}, starting by the furthest reachable parent,
	 * deduplicate tags that have a common key by favoring the value declared last (current tag(s) &gt; parent tag(s) &gt; grandparent tag(s))
	 * and return a {@link Map} of the deduplicated tags. Note that while the values are the "latest", the key iteration order reflects
	 * the tags' declaration order.
	 * <p>
	 * Tags can only be discovered until no parent can be inspected, which happens either
	 * when the source publisher has been reached or when a non-reactor intermediate operator
	 * is present in the parent chain (i.e. a stage that is not {@link Scannable} for {@link Attr#PARENT}).
	 *
	 * @return a {@link Map} of deduplicated tags from this {@link Scannable} and its reachable parents
	 * @see #tags()
	 */
	default Map<String, String> tagsDeduplicated() {
		...
	}
```
